src/amalgam/gen/rvv.c

```diff
@@ -214,18 +214,33 @@ void xnn_f32_rsum_ukernel__rvv_u4v(
   assert(output != NULL);
 
   batch >>= XNN_LOG2_SIZEOF_FLOAT;
-  int32_t n = __riscv_vsetvl_e32m4(batch);
-  vfloat64m1_t acc_f64v = __riscv_vfmv_s_f_f64m1(0.f, n);
-  do {
-    n = __riscv_vsetvl_e32m4(batch);
-
-    vfloat32m4_t in_f32v = __riscv_vle32_v_f32m4(input, n); input += n;
-    acc_f64v = __riscv_vfwredosum_vs_f32m4_f64m1(in_f32v, acc_f64v, n);
-    batch -= n;
-  } while (batch != 0);
-  vfloat64m1_t out_f64v = __riscv_vfmul_vf_f64m1(acc_f64v, params->scalar.scale, 1);
-  vfloat32mf2_t out_mf32v  = __riscv_vfncvt_f_f_w_f32mf2(out_f64v, 1);
-  __riscv_vse32_v_f32mf2(output, out_mf32v, 1);
+  int32_t n = __riscv_vsetvl_e32m1(batch);
+  vfloat32m1_t acc_f32v = __riscv_vfmv_s_f_f32m1(0.f, n);
+  n = __riscv_vsetvl_e32m4(batch);
+  int32_t n_sum = __riscv_vsetvl_e32m4(batch);
+  vfloat32m4_t sum_f32v = __riscv_vfmv_s_f_f32m4(0.f, n_sum);
+  if (batch >= n * 4) {
+    for (; batch >= n * 4; batch -= n * 4) {
+      vfloat32m4_t in0_f32v = __riscv_vle32_v_f32m4(input, n); input += n;
+      vfloat32m4_t in1_f32v = __riscv_vle32_v_f32m4(input, n); input += n;
+      vfloat32m4_t in2_f32v = __riscv_vle32_v_f32m4(input, n); input += n;
+      vfloat32m4_t in3_f32v = __riscv_vle32_v_f32m4(input, n); input += n;
+      vfloat32m4_t sum0_f32v = __riscv_vfadd_vv_f32m4(in0_f32v, in1_f32v, n);
+      vfloat32m4_t sum1_f32v = __riscv_vfadd_vv_f32m4(in2_f32v, in3_f32v, n);
+      sum_f32v = __riscv_vfadd_vv_f32m4(sum_f32v, sum0_f32v, n);
+      sum_f32v = __riscv_vfadd_vv_f32m4(sum_f32v, sum1_f32v, n);
+    }
+  }
+  if (batch > 0) {
```

Comment:
If it is expected that the routine is called with small input sizes, which are performance critical, it is good to check for a small input size first, and exit early after the work is done. Probably more of an impact for the LMUL=1 case?

Reply:
Yeah, this probably doesn't need to be addressed now. Only if you know that the function is going to be called with small lengths